### ISSUE TITLE
Some help with the new nest integration?

### ISSUE THREAD

**Issue by hmspain** [bwssytems/ha-bridge#36]
I've read all your documentation, and carefully added the username and password options to the command string.

The nest configuration tab shows up, and I can generate URLs.  They just don't "work".  I have no idea what I'm doing, and could use a little help?

Simple Home/Away gives me a request error.

**Comment by bwssytems** [comment]
After you generated you URL's, did you add them down below with the add button? It is a two step process to build them. The other thing is did you tell the Amazon Echo to discover your devices?

**Comment by hmspain** [comment]
Yes, but the nest URLs look very generic.  I added them anyway, saw them in the configuration, and discovered devices.  I have a ton of these configured to control Vera scenes, so I know the drill :-).

I then tried the test buttons.  Either nothing happens, or the error I posted.  I must be missing something.

I'm really anxious to get this working; I've been waiting so long to integrate my Nest with Echo! :-)


Harrison Spain
hmspain@gmail.com

**Comment by hmspain** [comment]
Trying the new 1.3.6.  Simple URL entry to set the temperature.  Forget all devices, discover devices.  "Amazon, set nest to 74".

[qtp13029138-20] ERROR spark.webserver.MatcherFilter -
java.lang.NullPointerException
        at com.bwssystems.HABridge.hue.HueMulator.lambda$10(HueMulator.java:383)
        at spark.RouteImpl$1.handle(RouteImpl.java:58)
        at spark.webserver.MatcherFilter.doFilter(MatcherFilter.java:162)
        at spark.webserver.JettyHandler.doHandle(JettyHandler.java:61)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:189)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
        at org.eclipse.jetty.server.Server.handle(Server.java:517)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:302)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:245)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
        at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
        at java.lang.Thread.run(Thread.java:745)

**Comment by bwssytems** [comment]
Could you post the configuration details of this item? Thanks

**Comment by hmspain** [comment]
Certainly; other than the line invoking the habridge jar file, I'm not sure what you need.  I have many scenes configured along with a couple harmony hubs.  The nest is not configured at this point.

You talk about a new method for backing up your config; perhaps if you showed me where that was and I could Email or upload the file?

**Comment by bwssytems** [comment]
Could you post the details of the next item you are having issues with? Click edit on the item and then copy that information into a post here.

The backup is on the main page, scroll to the bottom and you'll see the back up item.

**Comment by bwssytems** [comment]
Also,dumb question here so don't shoot me.. You say you don't have the nest configured at this point... Do you have a Nest Thermostat?

**Comment by hmspain** [comment]
Yes, a Gen3 Nest thermostat :-).

**Comment by bwssytems** [comment]
Ok, there shouldn't be an issue if you use the buttons on the nest device screen to build the bridge device. I have it working just fine and others are working as well. So i'll need the device configuration to debug.

**Comment by hmspain** [comment]
Click on the item?  You mean the item that should control temperature?  OK.

I can upload the device.db if that is the config you are looking for.

The device looks like:

Name: Living Room Temperature
Target: default
Map Type: Nest Thermostat
Map ID: undefined-SetTemp
On URL: {"name":"undefined","control":"temp","temp":"${intensity.percent}"}
Off URL: {"name":"undefined","control":"temp","temp":"${intensity.percent}"}
Http Verb: ---Please select---

**Comment by hmspain** [comment]
> The backup is on the main page, scroll to the bottom and you'll see the back up item.

Your mean the "Home" tab?  There is no "main" page I can find.

At the bottom are "Bridge settings"

BTW; I'm using Chrome, so if I need to switch to IE, just let me know.

**Comment by bwssytems** [comment]
There it is, it has the name as undefined and that would not be correct as that si what is put in the field when the variable is not set. So what does list look like in your Nest config screen?

**Comment by bwssytems** [comment]
Ahh, what version are you running? I think you have a buggy version as V1.3.0 is broke for the Nest feature. The backup feature was added in V1.3.5 and the latest version is v1.3.7.

**Comment by hmspain** [comment]
I always update to the latest as soon as it is released; I'm running 1.3.7 :-).

**Comment by hmspain** [comment]
> There it is, it has the name as undefined and that would not be correct as that si what is put in the 
> field when the variable is not set. So what does list look like in your Nest config screen?

Good news, we found something.  I'll get you the Nest config screen data when I get home this afternoon.

**Comment by bwssytems** [comment]
you need to rebuild your nest items as if they were created before v1.3.1, they are incorrect as to what you listed above.

**Comment by hmspain** [comment]
Yup, I've been rebuilding with each version hoping for a different result.  The key is that name undefined.  Given that this is not the expected behavior, we can trace back and figure out why the generated URL is created that way.

I keep thinking my username and password are wrong, but I've checked multiple times, and can log into nest.com with the pair.  Hmmm.  What else might make the generated URLs show up undefined?

**Comment by hmspain** [comment]
When I click on the "Temp" button in the "Nest Items List" tab, I get:

Name: Living Room Temperature

On URL: {"name":"undefined","control":"temp","temp":"${intensity.percent}"}

Off URL: {"name":"undefined","control":"temp","temp":"${intensity.percent}"}

What should it look like?  Perhaps I can make the adjustment manually.

**Comment by hmspain** [comment]
habridge appears to "see" the Nest thermostat, so I think the java command line switches (with username/password) are working.  The Nest Items List contains two items "Living Room(011T)" and "Home".

**Comment by bwssytems** [comment]
Interesting, as "Home" Should be what 'undefined' is. I'll retest and see what is up.

**Comment by bwssytems** [comment]
Ok, I think I have an answer. When you go to home.nest.com and go to settings->Home Info, do you have "name" set? That is what is used by the bridge.

**Comment by hmspain** [comment]
Yes, the name (I must have set) in nest.com "Home Info" is simply "Home":

I tried all sorts of combinations last night, but I set the name to "Living Room Temperature", and "Living Room(011T)".  I got Echo to say OK to "Echo, set Living Room Temperature to 74", but the Nest never budged.

I'll put "Home" in for the name, and "Living Room Temperature(011T)" for the undefined.  Are those the only two fields we have to mess with?

**Comment by bwssytems** [comment]
Ok, Some more info as I was looking at this from the value "target:default" which should be "target:Home" in your case

So the name you see on the screen for a thermostat to configure is "Living Room (011T)" which means  the code is finding the location and the name. The name that needs to be used is not the Home Info name but the device id name that is in the data. Sorry I went down that path. So, the "name:undefined" should be something like "name:abcdefg011T" and the"mapid:undefined-SetTemp" should be  "mapid:abcdefg011T-SetTemp".

All these are built through the javascript. What browser are you using? As it seems we may have a caching issue on the browser.

**Comment by hmspain** [comment]
I always use Chrome; sorry!

**Comment by hmspain** [comment]
It seems like we are real close, I look forward to trying that combination, it's not one I've tried before.

Could you also suggest the voice sequence?  Is this "Echo set the Living Room temperature to 74?" or "Excho turn on the Living Room to 74?" ???

**Comment by bwssytems** [comment]
Ok, I need you to execute this URL in your browser and then cut and paste the output to here. Thanks

Replace the appropriate items below with the bridge ip and the port you have it running on....
http://<YOUR BRIDGE IP>:<YOUR PORT>/api/devices/nest/items

**Comment by hmspain** [comment]
[{"name":"Home","id":"7fd7ac70-df64-11e1-a1c5-1231390e4661","type":"Home","location":"Los Alamitos, CA"},{"name":"Living Room(011T)","id":"09AA01AC3115011T","type":"Thermostat","location":"Living Room - Home"}]

**Comment by hmspain** [comment]
Argh, so close I can taste it :-).

Made all the changes (see below), discovered devices, and Echo says "OK", but the thermostat doesn't budge...

Target = default

Map ID = Living Room(011T)-SetTemp

On URL = {"name":"Living Room(011T)","control":"temp","temp":"${intensity.percent}"}

Off URL = {"name":"Living Room(011T)","control":"temp","temp":"${intensity.percent}"}

Http Verb = ---Please select---

**Comment by hmspain** [comment]
The Echo was confusing my Living Room lights with my Living Room Nest.  I renamed the nest configuration entry to "nest", forgot all devices (it was a mess), and did another discovery.

Still no love.

Sorry, the device nest is not responding, please checking it's connection and power supply.

**Comment by bwssytems** [comment]
So, all the info is coming from the bridge correctly. It seems you have a cache issue and chrome is hanging on to the Javascript from v1.3.0. You will need to clear all browser data for it to work properly. Edge, IE and firefox will execute the javascript and create your device. I would like you to try that to so that we can fix your system. For building the URL, you used the name above, it is the id that is used to find the items.

**Comment by hmspain** [comment]
I thought for sure I went over to IE to see if the nest URL (habridge builds) was populated properly.  In any case, I will try that first thing this afternoon.

If I call the action "Nest", what are the commands I would give Echo?

**Comment by bwssytems** [comment]
It is just like setting a dim on a light. So your line above should be good "Alexa, set nest to 75"

**Comment by hmspain** [comment]
I tried the new Edge (Windows 10) and IE (Windows 10).  The "Temp" button under the Nest tab still generates:

{"name":"undefined","control":"temp","temp":"${intensity.percent}"}

{"name":"undefined","control":"temp","temp":"${intensity.percent}"}

for the on and off URL.  Nothing is that easy LOL.

I manually adjusted the on and off URL for my "Nest" device to:

{"name":"Living Room(011T)","control":"temp","temp":"${intensity.percent}"}

{"name":"Living Room(011T)","control":"temp","temp":"${intensity.percent}"}

which is slightly different that your recommendation.  I tried both "name":"Living Room(011T)" and "name:Living Room(011T)".

"Alexa, set nest to 75"

Sorry the device nest is not responding, please check...

**Comment by hmspain** [comment]
Full disclosure: I have habridge running on my QNAP NAS.  It works fine with all habridge "devices" except nest.

Since I'm running on my QNAP, port 8080 is being used. so I'm running habridge on port 8081.

**Comment by hmspain** [comment]
I still can't see the "backup" feature, and all that displays under bridge settings (for nest) is "nest.configured true".

**Comment by hmspain** [comment]
My "About" says habridge 1.3.0 even though I'm explicitly running 1.3.7.  Is the help about perhaps wrong or is something else going on?  I understand you had some issues with 1.3.0.

**Comment by hmspain** [comment]
Found it!  The 1.3.0 led me to the solution.

It appears there was a process running on port 8081 (from earlier attempts to get it running on my QNAP).

Port 8081 was/is responding with 1.3.0 code somehow.

I checked my habridge.sh file, and sure enough, it was starting the bridge on port 
*8082*.

Changed my URL to point to 8082 and the correct habridge version comes up.  The nest URL now generates properly.

{"name":"09AA01AC3115011T","control":"temp","temp":"${intensity.percent}"}

[strange, but I'll take it]

and best of all... it works!  Thank you for hanging in there with me!
